tree/dataframe/src/RDFColumnRegister.cxx

```diff
@@ -19,6 +22,54 @@ namespace ROOT {
 namespace Internal {
 namespace RDF {
 
+RDefineAndReaders::RDefineAndReaders(std::shared_ptr<RDefineBase> define, unsigned int nSlots)
+   : fDefine(std::move(define)), fReadersPerVariation(nSlots)
+{
+   assert(fDefine != nullptr);
+}
+
+std::shared_ptr<RDFInternal::RDefineReader>
+RDefineAndReaders::GetReader(unsigned int slot, const std::string &variationName, const std::type_info &tid)
+{
+   auto &defineReaders = fReadersPerVariation[slot];
+
+   auto it = defineReaders.find(variationName);
+   if (it != defineReaders.end())
+      return it->second;
+
+   auto *define = fDefine.get();
+   if (variationName != "nominal")
+      define = &define->GetVariedDefine(variationName);
+
+   auto insertion = defineReaders.insert({variationName, std::make_shared<RDefineReader>(slot, *define, tid)});
+   return insertion.first->second;
+}
+
+RVariationAndReaders::RVariationAndReaders(std::shared_ptr<RVariationBase> variation, unsigned int nSlots)
+   : fVariation(std::move(variation)), fReadersPerVariation(nSlots)
+{
+   assert(fVariation != nullptr);
+}
+
+////////////////////////////////////////////////////////////////////////////
+/// Return a column reader for the given slot, column and variation, or a nullptr if not available.
+std::shared_ptr<RVariationReader> RVariationAndReaders::GetReader(unsigned int slot, const std::string &colName,
+                                                                  const std::string &variationName,
+                                                                  const std::type_info &tid)
+{
+   assert(IsStrInVec(variationName, fVariation->GetVariationNames())); // otherwise we should not be here
+
+   auto &varReaders = fReadersPerVariation[slot];
+
+   auto it = varReaders.find(variationName);
+   if (it != varReaders.end())
+      return it->second;
+
+   auto insertion = varReaders.insert(
```

Comment:
as far as I can tell it should not make a difference in terms of performance here (and `insert` is preferred because it has less template magic -> better error messages and compile times)